sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp

```diff
@@ -327,121 +329,335 @@ bool WinHttpTransport::VerifyCertificatesInChain(
+  void WinHttpAction::CompleteActionWithError(DWORD_PTR stowedErrorInformation, DWORD stowedError)
   {
-    // If an exception is thrown in the handler, log the error and terminate the connection.
-    Log::Write(Logger::Level::Error, "Exception Thrown: " + std::string(ex.what()));
+    // Note that the order of scope_exit and lock is important - this ensures that scope_exit is
+    // destroyed *after* lock is destroyed, ensuring that the event is not set to the signalled
+    // state before the lock is released.
+    auto scope_exit{m_actionCompleteEvent.SetEvent_scope_exit()};
+    std::unique_lock<std::mutex> lock(m_actionCompleteMutex);
+    m_stowedErrorInformation = stowedErrorInformation;
+    m_stowedError = stowedError;
   }
-}
 
-/**
- * @brief HTTP Callback to enable private certificate checks.
- *
- * This method is called by WinHTTP when a certificate is received. This method is called multiple
- * times based on the state of the TLS connection. We are only interested in
- * WINHTTP_CALLBACK_STATUS_SENDING_REQUEST, which is called during the TLS handshake.
- *
- * When called, we verify that the certificate chain sent from the server contains the certificate
- * the HTTP client was configured with. If it is, we accept the connection, if it is not,
- * we abort the connection, closing the incoming request handle.
- */
-void WinHttpTransport::OnHttpStatusOperation(HINTERNET hInternet, DWORD dwInternetStatus)
-{
-  if (dwInternetStatus != WINHTTP_CALLBACK_STATUS_SENDING_REQUEST)
+  DWORD WinHttpAction::GetStowedError()
+  {
+    std::unique_lock<std::mutex> lock(m_actionCompleteMutex);
+    return m_stowedError;
+  }
+  DWORD_PTR WinHttpAction::GetStowedErrorInformation()
+  {
+    std::unique_lock<std::mutex> lock(m_actionCompleteMutex);
+    return m_stowedErrorInformation;
+  }
+  DWORD WinHttpAction::GetBytesAvailable()
+  {
+    std::unique_lock<std::mutex> lock(m_actionCompleteMutex);
+    return m_bytesAvailable;
+  }
+
+  /**
+   * Called by WinHTTP when sending a request to the server. This callback allows us to inspect
+   * the TLS certificate before sending it to the server.
+   */
+  void WinHttpAction::StatusCallback(
+      HINTERNET hInternet,
+      DWORD_PTR dwContext,
+      DWORD internetStatus,
+      LPVOID statusInformation,
+      DWORD statusInformationLength) noexcept
   {
-    if (dwInternetStatus == WINHTTP_CALLBACK_STATUS_SECURE_FAILURE)
+    // If we're called before our context has been set (on Open and Close callbacks), ignore the
+    // status callback.
+    if (dwContext == 0)
     {
-      Log::Write(Logger::Level::Error, "Security failure. :(");
+      return;
     }
-    // Silently ignore if there's any statuses we get that we can't handle
-    return;
-  }
 
-  // We will only set the Status callback if a root certificate has been set.
-  AZURE_ASSERT(!m_options.ExpectedTlsRootCertificates.empty());
+    try
+    {
+      WinHttpAction* httpAction = reinterpret_cast<WinHttpAction*>(dwContext);
```

Comment:
Is this cast safe to do?

Reply:
I sure hope so. It's only unsafe if someone changes the code in `WinHttpSendRequest` to something other than the current action.